### PR TITLE
Feat functions, callables, return

### PR DESCRIPTION
**Pull request by mcmcgrath13**
### Implement functions as per [chapter 10 of Crafting Interpreters](https://craftinginterpreters.com/functions.html) with lots of translation to rust
* move `LoxValue` and related `impls` to `types.rs`
* add the `Function(UserFunction)` and `Builtin(NativeFunction)` `LoxValue` enum variants
* add the `Callable` trait and implement it for `LoxValue` (delegates to callable variants or errors), `UserFunction`, and `NativeFunction`
  * `Callable` adds the `call` method as well as some helpers for checking arity matches and reporting errors
* add the `UserFunction` struct, which takes a function declaration and a closure and allows it to be checked for validity and called
* add the `NativeFunction` struct, which takes an arity (for checkability) and a closure (lambda) an allows it to be called
* add AST nodes, parsing ,and interpreting for function declarations, function calls, and return statements
* add globals dictionary (though it's actually currently unused and I think it'll be revealed why it's needed in a future chapter I hope... or I'll delete it at the end)

### CODE REVIEW

File: src/interpreter.rs [hunk 1]
```diff
@@ -23,57 +16,38 @@ fn is_truthy(val: &LoxValue) -> bool {
...
 impl InterpreterError {
+    pub fn new(line: usize, location: impl AsRef<str>, message: impl AsRef<str>) -> Self {
+        Self {
+            line,
+            location: location.as_ref().to_string(),
+            message: message.as_ref().to_string(),
+            return_value: None,
+        }
+    }
+
```
Comment (mcastorina) [inline src/interpreter.rs]: I like the generic `impl AsRef<str>` here! I'm not too familiar with `impl Trait`.. Is that equivalent to doing `pub fn new<T: AsRef<str>>(...)`? I sort of recall that it's not fully monomorphic and can only be used with one type, but I could be wrong.

File: src/ast.rs [hunk 2]
```diff
@@ -88,6 +112,23 @@ impl PrettyPrinting for Stmt {
...
+                for parameter in parameters {
+                    s += " ";
+                    s += &parameter.print();
+                }
+                s += "(body";
+                for statement in body {
+                    s += " ";
+                    s += &statement.print();
+                }
+                s + "))"
```
Comment (mcastorina) [inline src/ast.rs]: Here's an alternative implementation idea:

```rust
format!(
    "({name} {parameters}(body {statements}))",
    name = name,
    parameters = parameters.iter().map(PrettyPrinting::print).collect::<Vec<_>>().join(" "),
    statements = body.iter().map(PrettyPrinting::print).collect::<Vec<_>>().join(" "),
)
```

I think it's equivalent, although the `iter().map().collect().join()` is not the most pretty. You could also implement `PrettyPrint` for `Vec<T: PrettyPrint>`:

```rust
impl<T> PrettyPrinting for Vec<T>
where
    T: PrettyPrinting,
{
    fn print(&self) -> String {
        self.iter()
            .map(PrettyPrinting::print)
            .collect::<Vec<_>>()
            .join(" ")
    }
}
```

which would make the usage:

```rust
format!(
    "({name} {parameters}(body {statements}))",
    name = name,
    parameters = parameters.print(),
    statements = body.print(),
)
```

Just spitballing alternative approaches.. feel free to ignore :)

File: src/interpreter.rs (hunk 1 above)
Comment (mcastorina) [reply src/interpreter.rs]: I [looked it up](https://stackoverflow.com/questions/47514930/what-are-the-differences-between-an-impl-trait-argument-and-generic-function-par) and they appear to be equivalent except for being able to specify the type.

```rust
fn foo<T: Trait>(t: T)
fn bar(t: impl Trait)

foo::<u32>(0) // this is allowed
bar::<u32>(0) // this is not
```

File: src/ast.rs (hunk 2 above)
Comment (mcmcgrath13) [reply src/ast.rs]: this is so much cleaner! thanks for the tip!

**Pull request merged**